lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java

```diff
@@ -16,28 +16,203 @@
 
 package com.palantir.lock.client.timestampleases;
 
+import com.github.rholder.retry.Attempt;
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.palantir.atlasdb.timelock.api.ConjureLockToken;
+import com.palantir.atlasdb.timelock.api.LeaseGuarantee;
+import com.palantir.atlasdb.timelock.api.LeaseIdentifier;
+import com.palantir.atlasdb.timelock.api.NamespaceTimestampLeaseRequest;
+import com.palantir.atlasdb.timelock.api.RequestId;
 import com.palantir.atlasdb.timelock.api.TimestampLeaseName;
-import com.palantir.lock.client.TimeLockUnlocker;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseRequests;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseResponse;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseResponses;
+import com.palantir.common.exception.AtlasDbDependencyException;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.lock.ConjureTimestampRangeTimestampSupplier;
+import com.palantir.lock.LimitingLongSupplier;
+import com.palantir.lock.client.LeasedLockToken;
+import com.palantir.lock.v2.TimestampLeaseResult;
 import com.palantir.lock.v2.TimestampLeaseResults;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.tritium.ids.UniqueIds;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import org.immutables.value.Value;
+import org.jetbrains.annotations.VisibleForTesting;
 
 final class TimestampLeaseAcquirerImpl implements TimestampLeaseAcquirer {
+    private static final SafeLogger log = SafeLoggerFactory.get(TimestampLeaseAcquirerImpl.class);
+
     private final NamespacedTimestampLeaseService delegate;
-    private final TimeLockUnlocker unlocker;
+    private final Unlocker unlocker;
+    private final Supplier<UUID> uuidSupplier;
+
+    private final Retryer<RequestAndResponse> retryer = RetryerBuilder.<RequestAndResponse>newBuilder()
+            .retryIfResult(requestAndResponse -> !wasRequestFullyFulfilled(requestAndResponse))
```

Comment:
What if instead of doing this calculation AFTER you've calculated the response, instead inline wasRequestFullyFulfilled check INSIDE acquireLeases.

Then add some sort of field in RequestAndResponse (and probably rename that) to indicate whether ther esponse is valid, and then this #retryIfResult just checks a boolean or whatever. That to me reads better because doing side-effecting stuff in retryIfResult feels nasty

Reply:
Now re-writing this, it's enough to have an optional and have it being empty signal things being bad

Can of course wrap this into some `ResponseWithValidation`